src/finalfusion/compat/text.py

```diff
@@ -0,0 +1,149 @@
+
+    Returns
+    -------
+    embeddings : Embeddings
+        Embeddings from the input file. The resulting Embeddings will have a
+        SimpleVocab, NdArray and Norms.
+    """
+    with open(file) as inf:
+        try:
+            first = next(inf)
+        except StopIteration:
+            raise ValueError("Can't read from empty embeddings file.")
+        line = re.split(r'(?a)\s+', first.rstrip())
+        cols = len(line[1:])
+        rows = sum(1 for _ in inf) + 1
+        inf.seek(0)
+        return _load_text(inf, rows, cols)
+
+
+def write_text(file: Union[str, bytes, int, PathLike], embeddings: Embeddings):
+    """
+    Write embeddings in text format.
+
+    Embeddings are un-normalized before serialization, if norms are present, each embedding is
+    scaled by the associated norm.
+
+    The output consists of utf-8 encoded lines with:
+        * word at the start of the line
+        * followed by whitespace
+        * followed by whitespace separated vector components
+
+    Parameters
+    ----------
+    file : str, bytes, int, PathLike
+        Output file
+    embeddings : Embeddings
+        Embeddings to write
+    """
+    _write_text(file, embeddings, False)
+
+
+def write_text_dims(file: Union[str, bytes, int, PathLike],
+                    embeddings: Embeddings):
+    """
+    Write embeddings in text-dims format.
+
+    Embeddings are un-normalized before serialization, if norms are present, each embedding is
+    scaled by the associated norm.
+
+    The output consists of utf-8 encoded lines with:
+        * `rows cols` on the **first** line
+        * word at the start of the line
+        * followed by whitespace
+        * followed by whitespace separated vector components
+
+    Parameters
+    ----------
+    file : str, bytes, int, PathLike
+        Output file
+    embeddings : Embeddings
+        Embeddings to write
+    """
+    _write_text(file, embeddings, True)
+
+
+def _load_text(file: TextIO, rows: int, cols: int) -> Embeddings:
+    words = []
+    matrix = np.zeros((rows, cols), dtype=np.float32)
+    for row, line in zip(matrix, file):
+        parts = re.split(r'(?a)\s+', line.rstrip())
+        words.append(parts[0])
+        row[:] = parts[1:]
+    storage = NdArray(matrix)
+    return Embeddings(storage=storage,
+                      norms=_normalize_ndarray_storage(storage),
+                      vocab=SimpleVocab(words))
+
+
+def _write_text(file: Union[str, bytes, int, PathLike], embeddings: Embeddings,
+                dims: bool):
+    vocab = embeddings.vocab
+    matrix = embeddings.storage[:len(vocab)]
+    with open(file, 'w') as outf:
+        if dims:
+            print(*matrix.shape, file=outf)
+        for idx, word in enumerate(vocab):
+            row = matrix[idx]
+            if embeddings.norms is not None:
+                row = row * embeddings.norms[idx]
+            print(word, ' '.join(map(str, row)), sep='\t', file=outf)
```

Comment:
I know tabs have the benefit that you can represent tokens with spaces, but IIRC word2vec when used in text mode uses spaces. So, other software may fail if there are suddenly tabs.

Maybe add a `wordSep` argument with a default of ` `. If a token cannot be properly be represented because it contains `wordSep`, it could raise an exception. Then advanced users who have a space in their embeddings could use `\t` as a separator.

Reply:
Hm, I thought we did that in finalfusion-rust, I made it configurable and replaced the default with `" "`!

Reply:
Just a regular space in ff-rust:

https://github.com/finalfusion/finalfusion-rust/blob/c056986d02523a9255ee8136051001eaab2feff2/src/compat/text.rs#L308

Reply:
Yup, misremembered that!